cpp/src/arrow/engine/substrait/expression_internal.cc

```diff
@@ -0,0 +1,902 @@
+            }
+            break;
+          }
+          default:
+            // Unimplemented construct, break out of loop
+            out.reset();
+            ref = nullptr;
+        }
+      }
+      if (out) {
+        return *std::move(out);
+      }
+      break;
+    }
+
+    case substrait::Expression::kIfThen: {
+      const auto& if_then = expr.if_then();
+      if (!if_then.has_else_()) break;
+      if (if_then.ifs_size() == 0) break;
+
+      if (if_then.ifs_size() == 1) {
+        ARROW_ASSIGN_OR_RAISE(auto if_, FromProto(if_then.ifs(0).if_(), ext_set));
+        ARROW_ASSIGN_OR_RAISE(auto then, FromProto(if_then.ifs(0).then(), ext_set));
+        ARROW_ASSIGN_OR_RAISE(auto else_, FromProto(if_then.else_(), ext_set));
+        return compute::call("if_else",
+                             {std::move(if_), std::move(then), std::move(else_)});
+      }
+
+      std::vector<compute::Expression> conditions, args;
+      std::vector<std::string> condition_names;
+      conditions.reserve(if_then.ifs_size());
+      condition_names.reserve(if_then.ifs_size());
+      size_t name_counter = 0;
+      args.reserve(if_then.ifs_size() + 2);
+      args.emplace_back();
+      for (auto if_ : if_then.ifs()) {
+        ARROW_ASSIGN_OR_RAISE(auto compute_if, FromProto(if_.if_(), ext_set));
+        ARROW_ASSIGN_OR_RAISE(auto compute_then, FromProto(if_.then(), ext_set));
+        conditions.emplace_back(std::move(compute_if));
+        args.emplace_back(std::move(compute_then));
+        condition_names.emplace_back("cond" + std::to_string(++name_counter));
+      }
+      ARROW_ASSIGN_OR_RAISE(auto compute_else, FromProto(if_then.else_(), ext_set));
+      args.emplace_back(std::move(compute_else));
+      args[0] = compute::call("make_struct", std::move(conditions),
+                              compute::MakeStructOptions(condition_names));
+      return compute::call("case_when", std::move(args));
+    }
+
+    case substrait::Expression::kScalarFunction: {
+      const auto& scalar_fn = expr.scalar_function();
+
+      auto id = ext_set.function_ids()[scalar_fn.function_reference()];
+
+      std::vector<compute::Expression> arguments(scalar_fn.args_size());
+      for (int i = 0; i < scalar_fn.args_size(); ++i) {
+        ARROW_ASSIGN_OR_RAISE(arguments[i], FromProto(scalar_fn.args(i), ext_set));
+      }
+
+      return compute::call(id.name.to_string(), std::move(arguments));
+    }
+
+    default:
+      break;
+  }
+
+  return Status::NotImplemented("conversion to arrow::compute::Expression from ",
+                                expr.DebugString());
+}
+
+Result<Datum> FromProto(const substrait::Expression::Literal& lit,
+                        const ExtensionSet& ext_set) {
+  if (lit.nullable()) {
+    // FIXME not sure how this field should be interpreted and there's no way to round
+    // trip it through arrow
+    return Status::Invalid(
+        "Nullable Literals - Literal.nullable must be left at the default");
+  }
```

Comment:
Ben's general take on this was that FromProto and ToProto should be 100% round-trippable, and anything that loses information should be rejected. He figured it'd be better to do Substrait plan manipulation to make a plan compatible with and optimized for Arrow in separate logic (or, at least, I asked him w.r.t. the optimization case and this was his answer; I'm extrapolating the compatibility point from that and his desire for everything to be perfectly round-trippable for testing and separation of concerns).

Personally I'm inclined to agree with you, and think the better approach is to have FromProto do the best it reasonably can (i.e. without making the code more difficult than it needs to be to do any conversion at all) to consume whatever the producer throws at it and convert it to an efficient equivalent in Arrow, and just don't do round-trip testing for conversions with information loss. Because, like you say, producing Substrait from Arrow and round-tripping in general is only a development/testing aid, not (currently?) a use case, and even if it were, you'd want that to do the best it can at generating efficient Substrait for the Arrow plan rather than rejecting Arrow plans where metadata would be lost (like struct field names, I think there's currently a rejection for that too). I think things will become needlessly complex even for simple plans if you separate these things, and result in a lot of code duplication (and extra tree-walking and intermediate conversion results at runtime, for as far as that is significant compared to the actual query execution). If round-tripping is expected to become a use-case down the line (communication between Arrow compute engines on different nodes, maybe?) we can also just add a flag to the convertors to enable checks for things that wouldn't round-trip.

Reply:
I'm not sure I can accept the rationale here.  The Arrow consumer shouldn't reject a plan because it has elements that we don't care about in Arrow.  This case would probably be less egregious than the "hints" case in `CheckRelCommon` but I think both situations are probably going to be pretty common.

I could buy the "sanitize the plan first" argument if we had some kind of representation other than Substrait that made sense.

Hmm...we could add a "deserialization options" object with a "safe" flag or something like that which rejected plans that can't be round tripped only if that flag is true.  That would quickly become painful however as we'd have to propagate that options object to a lot of function calls (probably the best approach would be to make the extension set an "option" so that we still have only one piece of tramp data).

Given the situation I'll create a follow-up JIRA for this.  I don't think it will take long for producers like Ibis to start creating plans with hints and it would be a shame if we couldn't process them.

Reply:
ARROW-15540